### PR TITLE
bpo-30397: Add re.Pattern and re.Match.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Also add and update docstrings.
Change `__module__` of the `error`, `Pattern` and `Match` types.

https://bugs.python.org/issue30397

### CODE REVIEW

File: Modules/clinic/_sre.c.h [hunk 1]
```diff
@@ -690,7 +690,7 @@ PyDoc_STRVAR(_sre_SRE_Match_span__doc__,
 "span($self, group=0, /)\n"
 "--\n"
 "\n"
-"For MatchObject m, return the 2-tuple (m.start(group), m.end(group)).");
+"For match object m, return the 2-tuple (m.start(group), m.end(group)).");
```
Comment (gvanrossum) [inline Modules/clinic/_sre.c.h:685]: Shouldn't that be "Match object"?

**Review by gvanrossum — COMMENTED** [review]
Thanks for doing this. I guess it's too scary to make these classes generic (like their counterparts in typing.py) -- after all we don't do that for other ABCs either.

**Comment by ilevkivskyi** [comment]
> Thanks for doing this. I guess it's too scary to make these classes generic (like their counterparts in typing.py) -- after all we don't do that for other ABCs either.

Yes. I was thinking about using these new classes in ``__extra__`` attributes of ``typing`` types, like we do for ``collections`` ABCs, but there is some special casing for ``Pattern`` and ``Match`` in ``typing`` so that it is better to keep the ``typing`` types as is.

File: Modules/clinic/_sre.c.h (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/clinic/_sre.c.h:685]: I don't know. The term "match object" is widely used in the documentation (unlikely to just added name `Match`).

**Pull request merged**